src/robot/libdocpkg/java9builder.py

```diff
@@ -0,0 +1,150 @@
+                return value
+        return ''
+
+    def _initializers(self, elements, constructors):
+        inits = [self._keyword_doc(elements, constructor)
+                 for constructor in constructors]
+        if len(inits) == 1 and not inits[0].args:
+            return []
+        return inits
+
+    def _keywords(self, elements, methods):
+        return [self._keyword_doc(elements, method) for method in methods]
+
+    def _keyword_doc(self, elements, method):
+        doc, tags = utils.split_tags_from_doc(self._get_doc(elements, method))
+        return KeywordDoc(
+            name=utils.printable_name(method.getSimpleName().toString(),
+                                      code_style=True),
+            args=self._get_keyword_arguments(method),
+            doc=doc,
+            tags=tags
+        )
+
+    def _get_keyword_arguments(self, method):
+        params = method.getParameters()
+        if not params:
+            return []
+        names = [param.getSimpleName().toString() for param in params]
+        if self._is_varargs(params[-1]):
+            names[-1] = '*' + names[-1]
+        elif self._is_kwargs(params[-1]):
+            names[-1] = '**' + names[-1]
+            if len(params) > 1 and self._is_varargs(params[-2]):
+                names[-2] = '*' + names[-2]
+        return names
+
+    def _is_varargs(self, param):
+        return (param.asType().toString().startswith('java.util.List') or
+                (param.asType().getKind() == TypeKind.ARRAY and
+                 param.asType().getComponentType().getKind() !=
+                    TypeKind.ARRAY))
+
+    def _is_kwargs(self, param):
+        return param.asType().toString().startswith('java.util.Map')
+
+
+def JavaDocumentation(path):
+    # Need to get the instance with reflection, or we ending up trying to
+    # access an internal class not exported.
+    doctool = ToolProvider.getSystemDocumentationTool()
+    file_manager = Class.forName("javax.tools.DocumentationTool") \
+        .getMethod("getStandardFileManager", DiagnosticListener, Locale,
+                   Charset) \
+        .invoke(doctool, None, Locale.US, StandardCharsets.UTF_8)
+
+    compiler = ToolProvider.getSystemJavaCompiler()
+    task = compiler.getTask(None, None, None, None, None,
+                            file_manager.getJavaFileObjectsFromStrings([path]))
+
+    type_element = task.analyze().iterator().next()
+    elements = task.getElements()
+    members = elements.getAllMembers(type_element)
+    qf_name = type_element.getQualifiedName().toString()
+
+    fields = filter(lambda field:
+                    Modifier.PUBLIC in field.getModifiers(),
+                    ElementFilter.fieldsIn(members))
+
+    constructors = filter(lambda constructor:
+                          Modifier.PUBLIC in constructor.getModifiers(),
+                          ElementFilter.constructorsIn(members))
+
+    methods = filter(lambda member:
```

Comment:
We've typically haven't used empty lines inside functions. It works fine here now, but might be unnecessary if filters are replaced with list comprehensions.

Reply:
Awesome, thanks! I am not familiar with Python at all, but agree with your points. Will refine tomorrow. About the reflection, it failed as it returns a type that is not exposed from the module, and that is forbidden since Java10. Could be supressed in Java9. I think your trick is inlining the doctool, so Jython not associates it with it’s runtime type. Will give it a try.